core/src/main/scala/org/broadinstitute/dsde/rawls/entities/local/EntityStatisticsCacheSupport.scala

```diff
@@ -0,0 +1,145 @@
+package org.broadinstitute.dsde.rawls.entities.local
+
+import com.typesafe.scalalogging.LazyLogging
+import io.opencensus.trace.Span
+import org.broadinstitute.dsde.rawls.dataaccess.SlickDataSource
+import org.broadinstitute.dsde.rawls.dataaccess.slick.{DataAccess, ReadAction, ReadWriteAction}
+import org.broadinstitute.dsde.rawls.model.{AttributeName, EntityTypeMetadata, Workspace, WorkspaceFeatureFlag}
+import org.broadinstitute.dsde.rawls.util.OpenCensusDBIOUtils.{traceDBIOWithParent, traceReadOnlyDBIOWithParent}
+
+import java.sql.Timestamp
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+
+/**
+  * helper methods that deal with entity type metadata and its cache
+  */
+trait EntityStatisticsCacheSupport extends LazyLogging {
+
+  implicit protected val executionContext: ExecutionContext
+  protected val dataSource: SlickDataSource
+  val workspaceContext: Workspace
+
+  import dataSource.dataAccess.driver.api._
+
+  val FEATURE_ALWAYS_CACHE_TYPE_COUNTS = WorkspaceFeatureFlag("alwaysCacheTypeCounts")
+  val FEATURE_ALWAYS_CACHE_TYPE_ATTRIBUTES = WorkspaceFeatureFlag("alwaysCacheAttributes")
+
+  case class CacheFeatureFlags(alwaysCacheTypeCounts: Boolean, alwaysCacheAttributes: Boolean)
+
+  /** convenience method for querying and then assembling an entity type metadata response.
+    * if this method retrieves metadata directly from entities and attributes (i.e. not from cache),
+    * it will update the cache with the results it found.
+    * */
+  def calculateMetadataResponse(dataAccess: DataAccess,
+                                        countsFromCache: Boolean,
+                                        attributesFromCache: Boolean,
+                                        outerSpan: Span = null): ReadWriteAction[Map[String, EntityTypeMetadata]] = {
+    val typesAndCountsQ = typeCounts(dataAccess, countsFromCache, outerSpan)
+    val typesAndAttrsQ = typeAttributes(dataAccess, attributesFromCache, outerSpan)
+    traceDBIOWithParent[Map[String, EntityTypeMetadata]]("generateEntityMetadataMap", outerSpan) { innerSpan =>
+      dataAccess.entityQuery.generateEntityMetadataMap(typesAndCountsQ, typesAndAttrsQ).flatMap { metadata =>
+        // and opportunistically save, if we have bypassed cache for all components
+        val saveCacheAction = if (countsFromCache || attributesFromCache) {
```

Comment:
It looks like we could turn on one of the cache feature flags and not the other and that could result in not updating the cache that was not set to always use the cached value even though we went through the trouble of running the query?  Is that correct?  If so it's probably not worth worrying about because the background cache updater should pick things up anyhow, but it's a little bit of an inconsistency and one more wrinkle to include when thinking about this process.

Reply:
yes, your assessment is absolutely correct (and good catch reverse-engineering all that!). You already saw broadinstitute/firecloud-develop#2815, which should help reduce that length-of-staleness.

I'd prefer to save the intricacies of opportunistically updating half the cache for some other PR, when/if we realize we need it. I stared at it a bit here and didn't have a good solution for it. The big complexity I want to be careful about is that any code we add in this spot is already inside a ` dataSource.inTransaction` closure … so I want to make absolutely certain that any operations we add don't extend the db transaction longer than necessary.

TL;DR I think this is thorny and I'm putting it off

Reply:
Thinking about this a little--still assuming I've got the right interpretation--couldn't we just change the logical or to an and to make the code opportunistically save any time we've recalculated the attribute or entity count parts of the cache?

Reply:
if only one flag was set, it means we pulled half the metadata from potentially-outdated cache, and the other half from "real". If we then saved this result back to the cache, we'd be marking the cache as updated - even though half of it might be outdated. I see what you're getting at - updating half the cache is still progress - but I am wary about then declaring the cache as up-to-date. I'm sticking to my stance of believing this needs further thought, not wanting to change it inside this PR.